### PR TITLE
[GLUTEN-8304][CORE] Add an optimization rule to collapse nested get_json_object functions

### PR DESCRIPTION
**Pull request by KevinyhZou**
## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)

(Fixes: \#8304)

## How was this patch tested?
test by ut

### CODE REVIEW

**Comment by PHILO-HE** [comment]
@KevinyhZou, I think this optimization can also be applied to Velox backend. Can you move this rule to a common place? BTW, there seems no need to add a new config. Maybe, just make the optimization always enabled. Thanks!

**Comment by PHILO-HE** [comment]
Hi @WangGuangxin, I note there is a config called `spark.sql.collapseGetJsonObject.enabled` according to this [link](https://www.volcengine.com/docs/6491/1263264). Is that optimization implemented by your team similar to this pr?

**Comment by KevinyhZou** [comment]
I have moved the rule to `gluten-subtriat` module, and keep a config `spark.gluten.sql.rewrite.nestedGetJsonObject` with default value `false` to control this rule. Could you help review this @PHILO-HE

File: gluten-substrait/src/main/scala/org/apache/gluten/extension/columnar/RewriteNestedGetJsonObjectExpressionRule.scala [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+      plan.withNewChildren(children)
+  }
+
+  private def optimizeNestedFunctions(
+      expr: Expression,
+      path: String = "",
+      isNested: Boolean = false): Expression = {
+
+    def getPathLiteral(path: Expression): Option[String] = path match {
+      case l: Literal if l.dataType.isInstanceOf[StringType] =>
```
Comment (PHILO-HE) [inline gluten-substrait/src/main/scala/org/apache/gluten/extension/columnar/RewriteNestedGetJsonObjectExpressionRule.scala:66]: Do we need to check StringType? I think json path should be always StringType.

**Review by PHILO-HE — COMMENTED** [review]
Some comments. Could you inject this rule for velox backend?

**Comment by KevinyhZou** [comment]
The rule now is injected for velox backend. @PHILO-HE

File: gluten-substrait/src/main/scala/org/apache/gluten/extension/columnar/CollapseGetJsonObjectExpressionRule.scala [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+package org.apache.gluten.extension.columnar
+
+import org.apache.gluten.GlutenConfig
+
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.plans.logical.{Filter, LogicalPlan, Project}
+import org.apache.spark.sql.catalyst.rules.Rule
+
+case class CollapseGetJsonObjectExpressionRule(spark: SparkSession) extends Rule[LogicalPlan] {
```
Comment (jackylee-ch) [inline gluten-substrait/src/main/scala/org/apache/gluten/extension/columnar/CollapseGetJsonObjectExpressionRule.scala:26]: Consider adding a comment to explain the purpose of the class and explain possible corner cases.

File: backends-clickhouse/src/test/scala/org/apache/gluten/execution/GlutenFunctionValidateSuite.scala [hunk 3]
```diff
@@ -268,6 +272,85 @@ class GlutenFunctionValidateSuite extends GlutenClickHouseWholeStageTransformerS
...
+      }
+      runQueryAndCompare(
+        "select get_json_object(get_json_object(get_json_object(string_field1, " +
+          "'$.a'), '$.y'), '$.z') from json_test where int_field1 = 6") {
+        x => assert(checkGetJsonObjectPath(x, "$.a.y.z"))
+      }
+      runQueryAndCompare(
+        "select get_json_object(get_json_object(get_json_object(string_field1, '$.a')," +
+          " string_field1), '$.z') from json_test where int_field1 = 6",
+        noFallBack = false
```
Comment (PHILO-HE) [inline backends-clickhouse/src/test/scala/org/apache/gluten/execution/GlutenFunctionValidateSuite.scala:324]: Why `noFallback = false`? CH backend only supports constant json path?

Comment (KevinyhZou) [reply backends-clickhouse/src/test/scala/org/apache/gluten/execution/GlutenFunctionValidateSuite.scala:324]: Yes

**Review by PHILO-HE — APPROVED** [review]
